### PR TITLE
[Enhancement] Enable multiple segments apply at once to reduce disk IO for persistent index during apply

### PR DESCRIPTION
**Pull request by sevev**
To reduce the memory usage after we do a large data ingestion, we support applying by segment and we will flush the primary index in advance. However, apply by segment will increase the disk IO for persistent index even we have enough memory because when we update a segment, we may need to scan the persistent index file. 

This pr introduces support for batching segments in memory. When sufficient memory is available, it accumulates multiple segments and performs a batched update of the persistent index to reduce I/O operations on the persistent index during the apply process.

In my test env, one BE has one disk(HDD) and create a primary key table with one bucket, using tpcds data. 
|PrimaryKey Length| RowNum|  Update memory limit| Apply time(ms)|  Note |
|---------------------|-----------|--------------------|------------------------------|----|
|16 Bytes| 259196263 | 16 GB |  2014666 | branch-main |
|16 Bytes| 259196263 | 16 GB | 219978 | branch-opt |
|16 Bytes| 259196263 | 8 GB | 302050 | branch-opt |

When the memory is insufficient, an early flush will still be triggered, so the longest time-consuming is the same as before optimization.

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:

- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.1
  - [x] 3.0
  - [x] 2.5
  - [ ] 2.4

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 315 / 445 (70.79%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/storage/tablet_updates.cpp|195|318|61.32%|[1072, 1073, 1074, 1075, 1094, 1096, 1097, 1098, 1106, 1107, 1108, 1109, 1110, 1111, 1112, 1113, 1115, 1116, 1119, 1120, 1122, 1123, 1124, 1132, 1134, 1135, 1136, 1148, 1149, 1150, 1151, 1152, 1179, 1180, 1181, 1182, 1183, 1194, 1195, 1196, 1197, 1199, 1200, 1201, 1203, 1204, 1206, 1207, 1209, 1210, 1211, 1221, 1223, 1224, 1225, 1254, 1255, 1257, 1258, 1389, 1390, 1391, 1392, 1395, 1396, 1397, 1399, 1403, 1404, 1412, 1413, 1432, 1433, 1435, 1436, 1437, 1440, 1441, 1442, 1444, 1445, 1446, 1447, 1448, 1449, 1450, 1452, 1453, 1454, 1455, 1457, 1459, 1460, 1461, 1462, 1464, 1465, 1466, 1468, 1469, 1470, 1471, 1472, 1474, 1475, 1476, 1478, 1479, 1481, 1482, 1483, 1485, 1486, 1488, 1489, 1490, 1491, 1492, 1493, 1494, 1495, 1496, 1922]|
|:large_blue_circle:|src/storage/primary_index.cpp|110|117|94.02%|[649, 654, 656, 1269, 1270, 1311, 1344]|
|:large_blue_circle:|src/storage/primary_index.h|10|10|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 311 / 441 (70.52%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/storage/tablet_updates.cpp|195|318|61.32%|[1080, 1081, 1082, 1083, 1102, 1104, 1105, 1106, 1114, 1115, 1116, 1117, 1118, 1119, 1120, 1121, 1123, 1124, 1127, 1128, 1130, 1131, 1132, 1140, 1142, 1143, 1144, 1156, 1157, 1158, 1159, 1160, 1187, 1188, 1189, 1190, 1191, 1202, 1203, 1204, 1205, 1207, 1208, 1209, 1211, 1212, 1214, 1215, 1217, 1218, 1219, 1229, 1231, 1232, 1233, 1262, 1263, 1265, 1266, 1397, 1398, 1399, 1400, 1403, 1404, 1405, 1407, 1411, 1412, 1420, 1421, 1440, 1441, 1443, 1444, 1445, 1448, 1449, 1450, 1452, 1453, 1454, 1455, 1456, 1457, 1458, 1460, 1461, 1462, 1463, 1465, 1467, 1468, 1469, 1470, 1472, 1473, 1474, 1476, 1477, 1478, 1479, 1480, 1482, 1483, 1484, 1486, 1487, 1489, 1490, 1491, 1493, 1494, 1496, 1497, 1498, 1499, 1500, 1501, 1502, 1503, 1504, 1930]|
|:large_blue_circle:|src/storage/primary_index.cpp|106|113|93.81%|[649, 654, 656, 1265, 1266, 1307, 1340]|
|:large_blue_circle:|src/storage/primary_index.h|10|10|100.00%|[]|

File: be/src/storage/primary_index.cpp [hunk 1]
```diff
@@ -1208,6 +1242,38 @@ Status PrimaryIndex::_upsert_into_persistent_index(uint32_t rssid, uint32_t rowi
...
+    Status st;
+    uint32_t n = info.total_update_rows();
+    std::vector<Slice> keys;
+    std::vector<uint64_t> values;
+    values.reserve(n);
+    std::vector<uint64_t> old_values(n, NullIndexValue);
+    const Slice* vkeys = _build_persistent_keys(info._pk_columns, info._idx_range, &keys);
+    RETURN_IF_ERROR(_build_persistent_values(rowset_id, info, &values));
+    RETURN_IF_ERROR(_persistent_index->upsert(n, vkeys, reinterpret_cast<IndexValue*>(values.data()),
+                                              reinterpret_cast<IndexValue*>(old_values.data()), stat));
```
Comment (decster) [inline be/src/storage/primary_index.cpp:1256]: what if there are duplicate keys in vkeys.

File: be/src/storage/primary_index.cpp [hunk 2]
```diff
@@ -1173,6 +1185,28 @@ const Slice* PrimaryIndex::_build_persistent_keys(const Column& pks, uint32_t id
...
+            }
+        } else {
+            uint32_t idx_begin = idx_ranges[i].first;
+            const uint8_t* keys = key_cols[i]->raw_data() + idx_begin * _key_size;
+            for (size_t j = idx_ranges[i].first; j < idx_ranges[i].second; j++) {
+                key_slices->emplace_back(keys, _key_size);
+                keys += _key_size;
+            }
+        }
+    }
```
Comment (decster) [inline be/src/storage/primary_index.cpp:1206]: if there are 10 rows, 3 key columns, there will be 30 key_slices? how can it be used in persistent_index insert/update methods?

Comment (sevev) [reply be/src/storage/primary_index.cpp:1206]: > if there are 10 rows, 3 key columns, there will be 30 key_slices? how can it be used in persistent_index insert/update methods?

No, autually the `key_cols` means segments. For each segment, we will combine all key columns and encoded them as a new column and one `key_col` means one segment.

File: be/src/storage/primary_index.cpp [hunk 3]
```diff
@@ -359,6 +359,7 @@ class FixSliceHashIndex : public HashIndex {
                         LOG(ERROR) << "found duplicate in upsert data rssid:" << rssid << " key=" << keys[i].to_string()
                                    << " [" << hexdump(keys[i].data, keys[i].size) << "]";
                     }
+                    LOG(INFO) << "erase old kv, rssid:" << (old >> 32) << ", rowid:" << (uint32_t)(old & ROWID_MASK);
```
[the remark this answers is not in the document]
Comment (sevev) [reply be/src/storage/primary_index.cpp:362]: > debug log?

Yes, I will change it.

File: be/src/storage/primary_index.cpp (hunk 1 above)
Comment (sevev) [reply be/src/storage/primary_index.cpp:1256]: > what if there are duplicate keys in vkeys.

We will first update these keys in l0 memory table and get the old value from historical l1 files. So if there are duplicate keys(key1) in vkeys, the first key1 will insert into l0 memory and we will mark it as not found in l0 and get old value from l1 files. When we upsert second key1, the first key1 is in l0 memory table and we will mark it as delete.

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 260 / 398 (65.33%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/storage/tablet_updates.cpp|200|320|62.50%|[1074, 1075, 1076, 1077, 1078, 1079, 1089, 1090, 1092, 1093, 1094, 1095, 1101, 1102, 1103, 1104, 1115, 1116, 1117, 1118, 1127, 1128, 1129, 1130, 1169, 1171, 1172, 1173, 1178, 1180, 1181, 1182, 1186, 1196, 1197, 1198, 1199, 1200, 1225, 1226, 1227, 1228, 1229, 1234, 1235, 1236, 1237, 1238, 1242, 1270, 1271, 1272, 1273, 1279, 1280, 1281, 1398, 1399, 1400, 1401, 1422, 1442, 1466, 1474, 1477, 1478, 1479, 1492, 1504, 1505, 1506, 1852, 1853, 1854, 1855, 1856, 1857, 1858, 1862, 1863, 1864, 1865, 1866, 1869, 1870, 1871, 1872, 1873, 1874, 1875, 1890, 1891, 1892, 1893, 1894, 1900, 1901, 1902, 1903, 1904, 1905, 1906, 1918, 1919, 1934, 1935, 1936, 1937, 1942, 1943, 1944, 1945, 1951, 1952, 1958, 1959, 1960, 1961, 1962, 1963]|
|:large_blue_circle:|src/storage/primary_index.cpp|60|78|76.92%|[649, 1201, 1202, 1245, 1246, 1247, 1248, 1250, 1262, 1270, 1300, 1307, 1309, 1310, 1311, 1312, 1314, 1316]|

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 265 / 415 (63.86%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/storage/tablet_updates.cpp|171|293|58.36%|[1081, 1082, 1083, 1084, 1103, 1105, 1106, 1107, 1115, 1116, 1117, 1118, 1119, 1120, 1121, 1122, 1124, 1125, 1128, 1129, 1131, 1132, 1133, 1141, 1143, 1144, 1145, 1157, 1158, 1159, 1160, 1161, 1188, 1189, 1190, 1191, 1192, 1203, 1204, 1205, 1206, 1208, 1209, 1210, 1212, 1213, 1215, 1216, 1218, 1219, 1220, 1230, 1232, 1233, 1234, 1263, 1264, 1266, 1267, 1398, 1399, 1400, 1401, 1404, 1405, 1406, 1408, 1412, 1413, 1421, 1422, 1441, 1442, 1444, 1445, 1446, 1449, 1450, 1451, 1453, 1454, 1455, 1456, 1457, 1458, 1459, 1461, 1462, 1463, 1464, 1466, 1468, 1469, 1470, 1471, 1473, 1474, 1475, 1477, 1478, 1479, 1480, 1481, 1483, 1484, 1485, 1487, 1488, 1490, 1491, 1492, 1494, 1495, 1497, 1498, 1499, 1500, 1501, 1502, 1503, 1504, 1505]|
|:large_blue_circle:|src/storage/primary_index.cpp|84|112|75.00%|[648, 653, 655, 1264, 1265, 1306, 1326, 1328, 1329, 1330, 1331, 1332, 1333, 1335, 1336, 1337, 1338, 1339, 1341, 1485, 1486, 1487, 1488, 1489, 1491, 1492, 1493, 1494]|
|:large_blue_circle:|src/storage/primary_index.h|10|10|100.00%|[]|

**Pull request closed without merging**